identity-iota/src/credential/validator.rs

```diff
@@ -1,197 +1,1305 @@
+  }
 
-    // Resolve the holder DID Document and validate the digital signature.
-    let holder_doc: DocumentValidation = self.validate_document(holder_url).await?;
+  // helper function to see the kind of error validation may yield
+  pub(crate) fn validate_presentation_internal<
+    'a,
+    T: Serialize,
+    S: Serialize,
+    I: IntoIterator<Item = &'a [ResolvedIotaDocument]>,
+  >(
+    &self,
+    presentation: &Presentation<T, S>,
+    options: &PresentationValidationOptions,
+    trusted_issuers: I,
+    resolved_holder_document: &ResolvedIotaDocument,
+    fail_fast: bool,
+  ) -> Result<(), AccumulatedPresentationValidationError> {
+    // first run some preliminary validation checks directly on the presentation
+    let preliminary_presentation_validation_errors: OneOrMany<ValidationError> =
+      Self::preliminaty_presentation_validation(presentation, fail_fast)
+        .err()
+        .unwrap_or_default();
+
+    let mut compound_error = AccumulatedPresentationValidationError {
+      presentation_validation_errors: preliminary_presentation_validation_errors.into_vec(),
+      credential_errors: BTreeMap::<usize, AccumulatedCredentialValidationError>::new(),
+    };
 
-    let mut credentials: Vec<CredentialValidation<U>> = Vec::new();
+    let AccumulatedPresentationValidationError {
+      ref mut presentation_validation_errors,
+      ref mut credential_errors,
+    } = compound_error;
+
+    // now check the holder's signature
+    if presentation_validation_errors.is_empty() || (!fail_fast) {
+      if let Err(proof_error) = Self::verify_presentation_signature(
+        presentation,
+        resolved_holder_document,
+        &options.presentation_verifier_options,
+      ) {
+        presentation_validation_errors.push(proof_error);
+      }
+    }
 
-    // Resolve and validate all associated credentials.
-    for credential in presentation.verifiable_credential.iter() {
-      // TODO: do we need to allow different VerifierOptions for each credential?
-      credentials.push(
-        self
-          .validate_credential(credential.clone(), VerifierOptions::default())
-          .await?,
+    // if any of the preliminary validations failed and fail_fast is true we must return now
+    if (!presentation_validation_errors.is_empty()) && fail_fast {
+      return Err(compound_error);
+    }
+
+    let mut trusted_issuers_iter = trusted_issuers.into_iter();
+    // validate the presentations credentials
+    for (position, credential) in presentation.verifiable_credential.iter().enumerate() {
+      let trusted_issuers = trusted_issuers_iter.next().expect(
+        "
+      incorrect parameters passed to private function validate_presentation_internal: 
+      the iterator over trusted issuers returns less elements than the number of credentials in the presentation. 
+      This is a bug which we encourage reporting at https://github.com/iotaledger/identity.rs/issues
+      ",
```

Comment:
This is an odd way of expecting issuers. I am still of the opinion that exact issuer-credential matching should be done manually since the criteria to do so varies per application. Basically we should just validate that all of the credentials in a presentation are signed by at least one of the trusted issuer documents.